// @better-auth/infra/dist/pow-BUuN_EKw.mjs
import "./constants-DdWGfvz1.mjs";
//#region src/dash-client.ts
function resolveDashUserId(input, options) {
	return input.userId || options?.resolveUserId?.({
		userId: input.userId,
		user: input.user,
		session: input.session
	}) || input.user?.id || input.session?.user?.id || void 0;
}
const dashClient = (options) => {
	return {
		id: "dash",
		getActions: ($fetch) => ({ dash: { getAuditLogs: async (input = {}) => {
			const userId = resolveDashUserId(input, options);
			return $fetch("/events/audit-logs", {
				method: "GET",
				query: {
					limit: input.limit,
					offset: input.offset,
					organizationId: input.organizationId,
					identifier: input.identifier,
					eventType: input.eventType,
					userId
				}
			});
		} } }),
		pathMethods: { "/events/audit-logs": "GET" }
	};
};
//#endregion
//#region src/crypto.ts
function randomBytes(length) {
	const bytes = new Uint8Array(length);
	crypto.getRandomValues(bytes);
	return bytes;
}
function bytesToHex(bytes) {
	return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
async function hash(message) {
	const msgBuffer = new TextEncoder().encode(message);
	const hashBuffer = await crypto.subtle.digest("SHA-256", msgBuffer);
	return bytesToHex(new Uint8Array(hashBuffer));
}
//#endregion
//#region src/abort-signal.ts
/**
* Returns an AbortSignal that aborts after `ms`.
* Uses `AbortSignal.timeout` when present; otherwise `AbortController` + `setTimeout`
* for React Native / older runtimes where `.timeout` is missing.
*/
function timeout(ms) {
	if (typeof AbortSignal.timeout === "function") return AbortSignal.timeout(ms);
	const controller = new AbortController();
	setTimeout(() => controller.abort(), ms);
	return controller.signal;
}
//#endregion
//#region src/identification-client.ts
function generateRequestId() {
	const hex = bytesToHex(randomBytes(16));
	return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20, 32)}`;
}
async function identify(baseURL, payload, signal) {
	const base = baseURL.replace(/\/$/, "");
	await fetch(`${base}/identify`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(payload),
		signal: signal ?? timeout(5e3)
	});
}
//#endregion
//#region src/sentinel/pow.ts
function hasLeadingZeroBits(hash, bits) {
	const fullHexChars = Math.floor(bits / 4);
	const remainingBits = bits % 4;
	for (let i = 0; i < fullHexChars; i++) if (hash[i] !== "0") return false;
	if (remainingBits > 0 && fullHexChars < hash.length) {
		if (parseInt(hash[fullHexChars], 16) > (1 << 4 - remainingBits) - 1) return false;
	}
	return true;
}
async function solvePoWChallenge(challenge) {
	const { nonce, difficulty } = challenge;
	let counter = 0;
	while (true) {
		if (hasLeadingZeroBits(await hash(`${nonce}:${counter}`), difficulty)) return {
			nonce,
			counter
		};
		counter++;
		if (counter % 1e3 === 0) await new Promise((resolve) => setTimeout(resolve, 0));
		if (counter > 1e8) throw new Error("PoW challenge took too long to solve");
	}
}
function decodeBase64ToUtf8(encoded) {
	if (typeof globalThis.atob === "function") return globalThis.atob(encoded);
	throw new Error("[Sentinel] Base64 decode requires atob (browser, Hermes, or Bun)");
}
function encodeUtf8ToBase64(str) {
	if (typeof globalThis.btoa === "function") return globalThis.btoa(str);
	const bytes = new TextEncoder().encode(str);
	const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
	let out = "";
	for (let i = 0; i < bytes.length; i += 3) {
		const b0 = bytes[i];
		const b1 = bytes[i + 1] ?? 0;
		const b2 = bytes[i + 2] ?? 0;
		const triple = b0 << 16 | b1 << 8 | b2;
		const pad = i + 2 >= bytes.length ? i + 1 >= bytes.length ? 2 : 1 : 0;
		out += chars[triple >> 18 & 63];
		out += chars[triple >> 12 & 63];
		out += pad < 2 ? chars[triple >> 6 & 63] : "=";
		out += pad < 1 ? chars[triple & 63] : "=";
	}
	return out;
}
function decodePoWChallenge(encoded) {
	try {
		const decoded = decodeBase64ToUtf8(encoded);
		return JSON.parse(decoded);
	} catch {
		return null;
	}
}
function encodePoWSolution(solution) {
	return encodeUtf8ToBase64(JSON.stringify(solution));
}
//#endregion
export { identify as a, dashClient as c, generateRequestId as i, encodePoWSolution as n, bytesToHex as o, solvePoWChallenge as r, hash as s, decodePoWChallenge as t };
